Corrected version detection for new pandas

PierreBdR opened the pull request:
Version 0.17 of pandas doesn't offer the 'version' module anymore. Instead, it provides the version via the default `__version__` attribute. I modified the `safe_version` function to allow for alternative means of finding the version for a given module. For example, to reliably get the version of pandas the call is:

    safe_version(pandas, ['version', 'version'], '__version__')

And now it works again.

bashtage commented:
Probably need to rebase on master now

PierreBdR commented:
@bashtage I'm not sure I understand ... I made a single commit from master, why do I need to rebase?

josef-pkt commented:
@PierreBdR   I just merged a PR that is supposed to fix this and other compatibility numpy and pandas issues. I haven't compared yet whether your PR here should still be merged..

PierreBdR commented:
@josef-pkt After checking I noticed a few differences:
1 - the other PR seems to have missed the `tools.print_version` module
2 - the other PR relies on the version string being 6 characters. Also, I don't really understand why, but it doesn't rely on the `strip_rc` function anymore. So unless I am missing something, we should use `strip_rc` and use the whole version number.
3 - It corrected quite a few other problems concerning changes of the pandas API.
I would say this PR still has some value (I rebased it on the new master)

josef-pkt commented:
Thanks for the update.

I'll try to figure out version issues again tomorrow. AFAIR, strip_rc should have problems with dev versions. We have another open PR to change the regular expression in setup.py, but so far none of the regular expressions handle every used case, AFAIK.

PierreBdR commented:
Ok, the version testing seems to be working (minus what strip_rc doesn't handle).

However, there are now a lot of errors due to a change in the numpy API. More precisely, they corrected an inconsistency between the interface of recarray and the documentation: while the `view` method always returned an ndarray, it now returns (correctly) a recarray. You can find the definition in the documentation as old as numpy 1.5, and before what happened if the type isn't specified wasn't defined: http://docs.scipy.org/doc/numpy-1.5.x/reference/generated/numpy.ndarray.view.html#numpy.ndarray.view 

Should I make a different PR for those or mix it with this one?

josef-pkt commented:
I didn't know that adding `type=ndarray' would fix this from our side
When I asked on the numpy mailing list, it was considered as a numpy bug
see #2668 for our issue which has the link to the numpy issue and that should have the link to the mailing list discussion.

I need to look at your solution to figure out what it does because I'm not familiar with the type argument for `view`.

The last remaining issues are in statespace #2672  where the reason for the failures and errors hasn't been found yet.

PierreBdR commented:
I've corrected a lot of the issues. Sadly it seems there are some more subtle ones. I will open issues on the tracker for them ...

rgommers commented:
@PierreBdR nice detective work! 

Do you understand the two ``test_representation`` failures? This PR is a major step forward, but it's tricky to merge it if the number of successful TravisCI builds goes from 4 to 0.

rgommers commented:
``strip_rc`` seems a bit pointless - it would strip off ``rc<int>`` but not ``b<int>`` (for beta releases). And it doesn't know anything about development versions (for pandas try ``strip_rc('0.17.0+githash')``). So just delete it I'd say.

bashtage commented on statsmodels/graphics/tsaplots.py:
```diff
@@ -201,10 +202,10 @@ def seasonal_plot(grouped_x, xticklabels, ylabel=None, ax=None):
-        if pandas.__version__ < '0.17.0':
-            df.sort()
-        else:
+        if is_pandas_min_version(StrictVersion('0.17.0')):
```
IMO this block is not the right way to handle this.  There should be a funciton in sm.compat.pandas called `sort_values` that would be something like

```
def sort_values(df, *args):
   if PD17:
       return df.sort_values(*args)
    else:
        return df.sort(*args)
```
There might be more mapping needed if kwargs changes location or names, although I don't think they did.  The point of `sm.compat` is to avoid any dependency-version-specific stuff in main code.

This also makes it easier in the long run since the function can be removed form `sm.compat.pandas` after the min version is >= 0.17 and then any imports of `sm.compat.pandas` can be removed.

PierreBdR replied:
To be honest, I think the proper way to solve this would be to create a `pandas_sort` function that does the sorting in a uniform way, independent from the version of pandas or numpy.

bashtage replied:
I agree - and this function should be in `sm.compat.pandas`.

rgommers replied:
Makes sense. I didn't realize that that was the policy.

PierreBdR replied:
Done. Although I've put the test outside the function definition ....

PierreBdR commented:
@rgommers  At least, I think I understand the problems with test_representations. AFAIK, they are due to corrections of the numpy API to match the documentation. I wrote a message to the PR #2672 explaining it. There are other issues though. But I don't think they are linked to my changes, because they are in parts of code I haven't touched in any way.

rgommers commented on statsmodels/tools/testing.py:
```diff
@@ -16,8 +16,10 @@ def strip_rc(version):
+    '''check whether pandas is at least min_version
+    '''
+    from pandas import __version__ as pversion
+    return StrictVersion(strip_rc(pversion)) >= min_version
```
This isn't right, will break on any dev version. Actually my version was wrong too, but at least it won't crash. Maybe use ``re.match(r'\d[.]\d+[.]\d+', pversion)`` plus check if there are any characters left afterwards (then it's an rc/dev/alpha/beta).

PierreBdR replied:
My take here is that we should have a better version of `StrictVersion` that can handle all that.
And `strip_rc` should be made useless for this. I'll try something tomorrow.

rgommers replied:
This may work out of the box: https://github.com/statsmodels/statsmodels/blob/master/statsmodels/compat/scipy.py#L20

The only difference between ``pandas`` and ``numpy`` is the dev version scheme: ``1.2.3.dev0+githash`` vs. ``1.2.3+githash``.

rgommers replied:
@jreback maybe now that you see this and that ``LooseVersion`` / ``StrictVersion`` are broken, you see the value in adding back ``version.short_version``? ``short_version`` is trivial to parse, while the class linked above is 130 LoC.

jreback replied:
not sure I see the problem
you seem to be jumping thru hoops to do simple version checking

we do this all the time (meaning using Loose/Strict Version) for lots of packages and it works fine

rgommers replied:
I pointed you to the complete PEP about this and gave you an example of why it's not fine yesterday.

jreback replied:
that's only relevant for a non released version (iow dev)

rgommers replied:
True. I'd think that it's a hard requirement to handle development versions of dependencies correctly in a project though. If you get test failures because your version checks say that ``0.17.0+githash == 0.17.0`` that would be quite annoying.

jreback replied:
never had that issue and we have quite a lot of deps in pandas

rgommers replied:
okay, guess I'm giving up. seems like you're not interested in people using pandas development versions in combination with statsmodels.

bashtage replied:
This is such a small issue, but what would be wrong with using 

```
def get_version(ver):
    """ Returns the numeric part of a version string, should work with PEP compliant versions"""
    valid = map(str,range(10)) + ['.']
    loc = 0
    for c in ver:
        if c not in valid:
            break
        loc += 1
    ver = ver[:loc]
    return ver[:-1] if ver[-1] == '.' else ver
```

jreback replied:
> okay, guess I'm giving up. seems like you're not interested in people using pandas development versions in combination with statsmodels.

@rgommers this attitude is quite puzzling

rgommers replied:
> this attitude is quite puzzling 

I don't want to offend you, or burn too much energy on this. But I'm pointing out a clear problem here, and I feel like you're simply ignoring my examples with "never seen it myself, so don't care". 

The actual code in this PR would simply error on a pandas dev version. I'd like pandas dev versions to work without a hitch - that's good for statsmodels users as well as for the quality of pandas releases.

jreback replied:
my point about the attitude was your way of checking is simply wrong. The following works just fine with and w/o dev versions.

```
In [6]: LooseVersion('0.17.0+126.gacf0eed') < '0.17.0'
Out[6]: False

In [7]: LooseVersion('0.17.0') < '0.17.0'
Out[7]: False

In [8]: LooseVersion('0.17.0') < '0.18.0'
Out[8]: True

In [9]: LooseVersion('0.17.0+126.gacf0eed') < '0.18.0'
Out[9]: True
```

rgommers replied:
> my point about the attitude was your way of checking is simply wrong.

Your very first example in the comment above is one that I gave to you as being *wrong*. I'll stop trying to clarify in my own words (apparently I'm not very good at it), but just copy from the PEP that talks about this (https://www.python.org/dev/peps/pep-0386/):

    The LooseVersion class is quite lax.
    ...
    This class makes any version string valid, and provides an algorithm to 
    sort them numerically then lexically.
    ...
    The problem with this is that while it allows expressing any nesting level it
    doesn't allow giving special meaning to versions (pre and post-releases
    as well as development versions), as expressed in requisites 2, 3 and 4.

ChadFulton commented:
Thanks for taking a look at this and for the PR! As you point out, this is a bit tricky.

Here are my thoughts. First, there are two issues at hand:

1. We ultimately need arrays to be column ordered (F-contiguous) for the Cython code, and they need to be in "wide" orientation (k_endog, nobs).
2. We don't want to copying arrays internally, if possible (i.e. if copying needs to happen, make the users do it, if possible).

Now, if the provided code is F-contiguous and wide, we have no problem because that's what we want. And if the provided code is C-contiguous and long, we have no problem because we can just transpose to get what we do want.

**MLEModel**

Statsmodels assumes that given data is in long format (nobs, k_endog), and the `MLEModel` class follows this convention. It assumes that anything that was passed to it is long format. It then explicitly casts to an ndarray and enforces / casts to C-contiguity, making a copy if necessary.

**Representation**

The job of the `Representation` class is to interface with the Cython library, so it is what needs the F-contiguous "wide" format.

This class uses memory contiguity as a proxy for dataset shape, and (assumed long) C-contiguous arrays are then transposed into (assumed wide) F-contiguous. Arrays which are neither F-contiguous or C-contiguous will always raise an error (but since most users interface with `Representation` through `MLEModel`, they can pass any type of array they like and it will be copied to contiguous memory if it isn't contiguous already).

Given the assumptions in `MLEModel` there is typically no problem. `MLEModel` forces a long C-contiguous array. Then it passes it to `Representation` which transposes it to get wide F-contiguous.

As you point out, the only problem is the 2-dim (but in terms of memory only 1-dim) case where C_CONTIGUOUS and F_CONTIGUOUS are both true; e.g. (1,10) and (10,1). Here contiguity can't tell us whether the array is intended as long or wide.

The current behavior is:

```python
if endog.flags['C_CONTIGUOUS']:
    endog = endog.T
```

And your fix is:

```python
if not endog.flags['F_CONTIGUOUS']:
    endog = endog.T
```

But I think this isn't quite enough, because within the `Representation` class we don't require input to be either long / wide and in this case we also don't get a clue from F / C contiguity. As an example, if an array of shape (1,10) is passed, the existing behavior will always run the conditional command and your proposed behavior will never run the conditional command. It may be nicer to have some sort of standard behavior which is alterable by an argument.

It seems to me that in the case that one of the dimensions has shape 1, the default should be to assume `k_endog=1`. I think of it as treating it like a 1-dim vector. So by default, a (10,1) would be assumed to be long format and a (1,10) would be assumed to be wide format. There is an optional `nobs` keyword argument which I think can be safely used to allow the non-default behavior to be specified.

Since the point of this line is to transpose from long to wide (and there is no effect on contiguity from transposing in the case we're considering), I think we can do:

```python
if endog.flags['C_CONTIGUOUS'] and (endog.shape[0] > 1 or nobs == 1):
    endog = endog.T
```

We would need to add to the `nobs` and / or `k_endog` documentation that the only way to truly get a dataset with nobs > 1 and k_endog > 1 is to pass the `nobs=1` argument.
And I think we would also need to alter `MLEModel.initialize_statespace` to pass `nobs=1` if indeed `nobs=1`.

I think we can save most of the unit tests here, by either changing their dimensions:

- `test_bind`: for the invalid tests, instead of (10,1) and (1,10), we can use (5,2) and (2,5) to get the right results. We may also want to add a couple of new lines for the now-new behavior that F-contiguous (10,1) will also have nobs=10, unless the nobs keyword was set, and similarly for the C-contiguous.
- `test_filter` and `test_predict` under the proposed behavior here, these could use either (10,1) or (1,10) data, but I think the existing (10,1) is more obvious.
- `test_mlemodel` you're right that we'll have to remove those lines, since obviously they will give different answers in different numpy versions.

Still not quite positive I've thought through all of the implications.

Does this sound like a reasonable strategy? Is there a better way out I haven't thought of?

If this is good, you are free to implement in this PR or else I am happy to implement in a separate PR.

Thanks!

bashtage commented:
> test_mlemodel you're right that we'll have to remove those lines, since obviously they will give different answers in different numpy versions.

Ideally there would be version specific versions of this test.  You can use something like 

```

@skipif(NPgt110)
def test_mlemodel():
    blah
```

variables like `NPgt110` should be defined in `sm.compat.numpy`

PierreBdR commented:
@ChadFulton I must say, although I understand where this is coming from, I am not a big fan of depending on C or F contiguity at all. But as this is the current behaviour ... changing it would probably too painful. Otherwise I agree with your proposal and will adapt my solution for it.

ChadFulton commented:
> I must say, although I understand where this is coming from, I am not a big fan of depending on C or F contiguity at all.

I think another solution would be nice too, but I can't think of one that works better here (of course I'm not an expert in numpy / pandas, so likely someone else could think of something). Ultimately it is just unavoidable that to pass the arrays to the Cython code, they must be F-contiguous and in wide format.

ChadFulton commented:
> Ideally there would be version specific versions of this test. You can use something like...

Good to know; sounds like something we should incorporate @PierreBdR.

PierreBdR commented:
Ok, I am now using LooseVersion to check pandas version ... I guess it will work better this way.
I also added a 'version' variable to `compat.pandas` with value `LooseVersion(pandas.__version__)` so to test for a specific version one can do:

    from statsmodels.compat.pandas import version as pandas_version
    if pandas_version >= '0.10.0':
        blah
    else:
        bloh

 I also changed the representation as suggested by @ChadFulton and it works. Part of the tests are only done for NumPy version < 0.10 (the 1D ones).

**Note:** Can someone look at the `test_dynamic_factor`? There is exactly one error because it seems that now the function doesn't converge quite correctly. But I couldn't find why the result changes when the version of pandas and/or numpy changes. Also, I don't know what the function is supposed to do ... so it doesn't help.

bashtage commented:
Should this be merged than an new PR can be worked on that tried to fix remaining issues?

josef-pkt commented:
If it's ready, then I'll merge.
I'm glad if I don't have to dig into the details. Thanks to all the participants here.

PierreBdR commented:
We cannot merge yet. There is another issue with `pandas.__version__`. For some reason I don't understand, it's not found on CI Travis, while I don't have any issues on my machine. I'll correct that and then I think we can merge this and work on the other issue in another PR.

bashtage commented on statsmodels/compat/pandas.py:
```diff
@@ -0,0 +1,11 @@
+import pandas
```
needs to be import pandas as pd

bashtage replied:
My suggestion didn't work.  Instead you can do this

in `compat.init.py`

```
from __future__ import absolute_import
from distutils.version import LooseVersion

from statsmodels.compat.python import *
import pandas as pd

pandas_version = LooseVersion(pd.__version__)
```
in `compat.pandas`
```
import statsmodels.compat

version = statsmodels.compat.pandas_version

if version >= '0.17.0':
    def sort_values(df, *args, **kwargs):
        return df.sort_values(*args, **kwargs)
else:
    def sort_values(df, *args, **kwargs):
        return df.sort(*args, **kwargs)
```

bashtage commented on statsmodels/compat/pandas.py:
```diff
@@ -0,0 +1,11 @@
+import pandas
+from distutils.version import LooseVersion
+
+version = LooseVersion(pandas.__version__)
```
needs to be pd.__version__

bashtage commented on statsmodels/graphics/tsaplots.py:
```diff
@@ -201,10 +201,7 @@ def seasonal_plot(grouped_x, xticklabels, ylabel=None, ax=None):
-            df.sort()
-        else:
-            df.sort_values(inplace=True)
+        df.sort(inplace=True)
```
Is this a series or a dataframe?  Isn't `compat.pandas.sort_values` needed here?

bashtage replied:
In old pandas this is a TimeSeries.  TimeSeries doesn't support inplace sorting, so just do

```
df = sorted_values(df)
```

bashtage replied:
That also doesn't work since TimeSeries sorted inplace by default, so need a change in `compat.pandas`

to be 
```
if version >= '0.17.0':
    def sort_values(df, *args, **kwargs):
        out = df.sort_values(*args, **kwargs)
        return out if out is not None else df
else:
    def sort_values(df, *args, **kwargs):
        out = df.sort(*args, **kwargs)
        return out if out is not None else df
```

I think this should handle all `inplace` combinations.

PierreBdR replied:
In newer pandas, `sort_values` *requires* a list of keys to sort (at least for data frames). Otherwise, we need to use `sort_index`.  when replacing `sort` with `sort_values` systematically, I had this issue in quite a few places.

But in any case, I am preparing an environment with old versions of numpy, scipy and pandas to check all these and correct them.

bashtage commented:
Ok, final comment -- all you really need is absolute_import.  The only change needed to for `compat.pandas` to be 

```
from __future__ import absolute_import

from distutils.version import LooseVersion

import pandas as pd

version = LooseVersion(pd.__version__)

if version >= '0.17.0':
    def sort_values(df, *args, **kwargs):
        out = df.sort_values(*args, **kwargs)
        return out if out is not None else df
else:
    def sort_values(df, *args, **kwargs):
        out = df.sort(*args, **kwargs)
        return out if out is not None else df
```

No need to change `compat/init.py`.  Also need to use `sorted_values(df)` inplace of `df.sort(inplace)`

_Note_: The above code doesn't really work correctly since it should return `None` when `inplace=True` -- this version will always return the object, although it does implement inplace sorting when asked.

PierreBdR commented:
Ok, so I hope the latest version corrects most issues. In the end I had to have three versions of sort_values:
 1 - version >= 0.17 uses df.sort_values
 2 - version >= 0.14 uses df.sort but always set inplace to False by default
 3 - version < 0.14 didn't have the 'inplace' argument to df.sort for non-DataFrame types, so I added it with False by default and copy the object if we don't want to do the sort in place

PierreBdR commented:
@bashtage For me the more pythonic code is the simpler, correct, code. I think what my code is doing is clear: get the argument with `False` for default value, and delete it. The `setdefault` is there to ensure the key is always there so I don't need to test for it.

In any case, now I have a weird behaviour: according to Travis, there are some tests failing with the latest version of pandas, numpy, scipy, ... This would be fine if I could reproduce this on my machine, but I cannot! Can anyone confirm they can reproduce the failure observed by Travis?

Also, I do get a failure, but here is the one I get:

```
======================================================================
FAIL: statsmodels.tsa.statespace.tests.test_dynamic_factor.TestDynamicFactor_ar2_errors.test_mle
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/barbier/prog/projects/statsmodels/statsmodels/tsa/statespace/tests/test_dynamic_factor.py", line 478, in test_mle
    assert_allclose(res.llf, self.results.llf, atol=1e-3)
  File "/home/barbier/local/py_latest/local/lib/python2.7/site-packages/numpy/testing/utils.py", line 1347, in assert_allclose
    verbose=verbose, header=header)
  File "/home/barbier/local/py_latest/local/lib/python2.7/site-packages/numpy/testing/utils.py", line 708, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0.001

(mismatch 100.0%)
 x: array(606.8223952852253)
 y: array(607.4201043440738)
```

The only difference I can see is that I am using Python 2.7.6 while Travis has Python 2.7.10.

PierreBdR commented:
@bashtage Sorry, I didn't answer your second comment. I don't understand what shouldn't be working? Won't old versions of pandas always sort in place? So if `inplace` is `False`, I should be able to copy the object, sort it and return it. Otherwise, sorting the object in place and returning it should be working too. Or am I getting something wrong?

josef-pkt commented:
I restarted the last python job on travisci because it had timed out.
Now it fails with conda command not found.

bashtage commented:
> Now it fails with conda command not found.

This is why:

```
PREFIX=/home/travis/miniconda2
```

For some reason installed to non-standard place.  Probably something about restarting when using the docker environment not cleaning up the original miniconda.  Should probably add a 

```
rm -rf /home/travis/miniconda
```

to `travis.yml` before running miniconda

bashtage commented:
@josef-pkt It isn't going to pass on Numpy 1.10 due to a few remaining issues in `tsa`.  While maybe not perfect, it fixes a lot of high level bug and so I think I would probably get this in.

ChadFulton commented:
>  It isn't going to pass on Numpy 1.10 due to a few remaining issues in tsa.

I am planning to take another look at those (I think they're mostly state space issues) after this PR is merged so as to eliminate the noise coming from the F / C ordering issues.

josef-pkt commented on statsmodels/tsa/statespace/tests/test_dynamic_factor.py:
```diff
@@ -473,8 +473,8 @@ def __init__(self):
-            res = mod.fit(method='lbfgs', maxiter=10000, disp=-1)
-            res = mod.fit(res.params, method='nm', maxiter=10000, maxfev=10000, disp=False)
+            res1 = mod.fit(method='lbfgs', maxiter=10000, disp=-1)
+            res = mod.fit(res1.params, method='nm', maxiter=10000, maxfev=10000, disp=False)
```
this is not required, keeps res in memory a bit longer.
doesn't matter

PierreBdR replied:
Yes sorry, I changed that when I was trying to understand where the error came from: from the computation of res or res1 ...

josef-pkt commented on statsmodels/tsa/statespace/tests/test_representation.py:
```diff
@@ -601,34 +602,47 @@ def empty_endog():
-    mod.bind(np.arange(10)*1.)
-    assert_equal(mod.nobs, 10)
+    mod.bind(np.arange(10).reshape((5,2))*1.)
+    assert_equal(mod.nobs, 5)
```
This changes the test case AFAICS, `# Test valid (nobs x 1) endogenous array`

Is this intended or a bugfix? @ChadFulton

PierreBdR replied:
I did that following a suggestion above. The issue is that with the new numpy, 1D array are both F and C contiguous ... so the following checks always fail. I changed the main tests to a 2D array, and then added back the 1D array tests if numpy version is less than 1.10.

josef-pkt replied:
Ok, thanks for the explanation.
I see now the difference in the flags between numpy 1.9 and 1.10.

josef-pkt commented:
It looks fine to me, browsing the code.
The only question I have is about changing the test case. Otherwise, looks fine to merge.

josef-pkt commented:
Now master also fails with conda not found, independent of this PR, after merging unrelated PRs, which passed the travis tests before, except for the issues here
https://travis-ci.org/statsmodels/statsmodels/jobs/89143029

josef-pkt commented:
The miniconda2 (added 2) prefix is also used in a standard test run in the old travis infrastructure.
Maybe something has changed with miniconda.sh ?

```
$ ./miniconda.sh -b
PREFIX=/home/travis/miniconda2
```
https://travis-ci.org/statsmodels/statsmodels/jobs/89144408

Is it possible to add both `miniconda` and `miniconda2` to the path so at least one of them is found?
Or does this raise an error if the directory doesn't exist?

josef-pkt commented:
merging, thanks all
a fix for conda not found is in #2692

josef-pkt commented:
After merging this (and fixing miniconda), travis hangs on the test suite https://travis-ci.org/statsmodels/statsmodels/jobs/89156803
no output received for 10 minutes
And it looks like we lost the dots in the test output ...................................... in the python with latest version.

I don't know yet whether it's significant but the testrun takes now longer on recent python versions, especially 3.4. I don't know whether it's a consequence of the new container based infrastructure.

bashtage commented:
There is a big need for a rationalizing the small number of tests that take a large percentage of time.  These should be labeled as slow and then only run in a separate job

josef-pkt commented:
I don't think that's really a problem for the Travis test runs. Usually the first result was available after 15 minutes or a bit longer.  For checking my own PRs during development, I always run selected unit tests which is in almost all cases very fast.

The problem we have been having for some time now is that the time the test run takes is very version dependent. But until now it was only python 2.7 with the old versions and occasionally python 2.6 that slowed down to hit the 50 minutes time limit.
I think that's a substantive issue that most likely the optimization doesn't work well in some cases, or that the underlying test case is fragile (intentionally or not).
Because we don't have detailed results on individual unit tests on Travis, I have no idea which unit tests might cause problems only in some cases.  This is separate from test that are always slow, like some tests in emplike or nonparametric.

bashtage commented:
A tiny number of tests (0.5%) take over 50% of the time:

```
Percent
of 
Total   Time    Test
------  -----   ----
11.96	95.092	statsmodels.regression.tests.test_lme.TestMixedLM.test_compare_numdiff
6.50	51.655	statsmodels.emplike.tests.test_regression.TestRegressionPowell.test_ci_beta0
3.63	28.853	statsmodels.emplike.tests.test_aft.Test_AFTModel.test_betaci
3.39	26.928	statsmodels.emplike.tests.test_regression.TestRegressionPowell.test_ci_beta2
2.80	22.222	statsmodels.emplike.tests.test_regression.TestRegressionPowell.test_ci_beta1
2.32	18.464	statsmodels.emplike.tests.test_regression.TestRegressionPowell.test_ci_beta3
1.89	15.039	statsmodels.stats.tests.test_corrpsd.Test_Factor.test_corr_nearest_factor_sparse
1.81	14.396	statsmodels.genmod.tests.test_gee.TestGEEOrdinalCovType.test_cov_type
1.73	13.76	statsmodels.regression.tests.test_lme.TestMixedLM.test_vcomp_formula
1.62	12.841	statsmodels.regression.tests.test_lme.TestMixedLM.test_profile_inference
1.59	12.675	statsmodels.genmod.tests.test_gee.test_plots
1.51	11.981	statsmodels.regression.tests.test_lme.TestMixedLM.test_formulas
1.45	11.55	statsmodels.tsa.statespace.tests.test_dynamic_factor.TestDynamicFactor_ar2_errors.test_mle
1.40	11.117	statsmodels.emplike.tests.test_regression.TestRegressionNM.test_ci_beta1
1.28	10.197	statsmodels.emplike.tests.test_regression.TestRegressionNM.test_ci_beta0
1.28	10.184	statsmodels.tsa.statespace.tests.test_structural.test_lltrend_cycle_seasonal_reg_ar1
1.19	9.443	statsmodels.emplike.tests.test_regression.TestRegressionNM.test_ci_beta3
1.09	8.685	statsmodels.graphics.tests.test_dotplot.test_all
1.06	8.42	statsmodels.genmod.tests.test_gee.TestGEE.test_ordinal
1.03	8.217	statsmodels.regression.tests.test_lme.TestMixedLM.test_regularized
1.00	7.964	statsmodels.genmod.tests.test_gee.TestGEEOrdinalCovType.test_wrapper
0.98	7.811	statsmodels.emplike.tests.test_descriptive.TestDescriptiveStatistics.test_ci_corr
0.96	7.598	statsmodels.stats.tests.test_multi.test_pvalcorrection_reject
0.87	6.911	statsmodels.emplike.tests.test_regression.TestRegressionNM.test_ci_beta2
0.70	5.535	statsmodels.genmod.tests.test_gee.TestGEE.test_nominal
0.64	5.078	statsmodels.tsa.statespace.tests.test_mlemodel.test_fit_misc
	
```

The first column adds to 55.67%.

PierreBdR commented:
Increasing the verbosity of the test, it seems that the following test is the one blocking: `statsmodels.tsa.statespace.tests.test_dynamic_factor.TestDynamicFactor_ar2_errors.test_mle`
On my machine it takes only 3s to execute, and fails due to an error in the result.
